properly fix S24_LE mode with pcm512x codec

HiassofT opened the pull request:
This reverts the pcm512x downstream changes introduced in commit 185ea05465aac8bf02a0d2b2f4289d42c72870b7 / PR #1152

The downstream pcm512x changes caused a regression, it broke normal use of the 24bit format with the codec, eg when using simple-audio-card.

The actual bug with 24bit playback is the incorrect usage of physical_width in various drivers in the downstream tree which causes 24bit data to be transmitted with 32 clock cycles. So it's not the pcm512x that needs fixing, it's the soundcard drivers.

I had a go at fixing all the pcm512x based soundcard drivers. Actually using the codec driver as it was intended to be used simplified them quite a bit as I could remove all those set_bclk_ratio calls (and in some cases hw_params / ops as all they did was call set_bclk_ratio).

I've successfully tested 16, 24 and 32bit playback with the IQaudio DAC+ and JustBoom DAC cards. But as I don't have any of the other pcm512x based cards here some more testing and feedback would be welcome - not sure if I got all the soundcard drivers right.

@hifiberry could you check if this works with your DAC+ and DAC pro cards? The non-pro case in your hifiberry_dacplus driver looked rather odd to me - setting bclk ratio to a fixed value of 64 should have broken 16-bit playback IMHO. The allo-boss driver which seems to be a clone of the dacplus driver uses 2 * physical_width in the non-master case (which would be your non-pro case) - which looks more sensible to me.

@iqaudio @shawaj @babuenir @clivem @allocom @miquel83blauw @jgrulich it would be great if you could test this with your pcm512x based cards as well.

HiassofT committed:
ASoC: pcm512x: revert downstream changes

This partially reverts commit 185ea05465aac8bf02a0d2b2f4289d42c72870b7
which was added by https://github.com/raspberrypi/linux/pull/1152

The downstream pcm512x changes caused a regression, it broke normal
use of the 24bit format with the codec, eg when using simple-audio-card.

The actual bug with 24bit playback is the incorrect usage
of physical_width in various drivers in the downstream tree
which causes 24bit data to be transmitted with 32 clock
cycles. So it's not the pcm512x that needs fixing, it's the
soundcard drivers.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: iqaudio-dac: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: allo-boss-dac: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: hifiberry_dacplus: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: allo-piano-dac: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Also remove hw_params and ops as they are no longer needed.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: dionaudio_loco-v2: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Also remove hw_params and ops as they are no longer needed.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: allo-piano-dac-plus: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: raspidac3: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Also remove hw_params as it's no longer needed.

Signed-off-by: Matthias Reichl <hias@horus.com>

HiassofT committed:
ASoC: justboom-dac: fix S24_LE format

Remove set_bclk_ratio call so 24-bit data is transmitted in
24 bclk cycles.

Also remove hw_params as it's no longer needed.

Signed-off-by: Matthias Reichl <hias@horus.com>

shawaj commented:
Will do. Copying in @francesco-vannini and @mmilann too

popcornmix commented:
Looks good to me.

HiassofT commented:
rebased onto current rpi-4.14.y tree and fixed commit messages (S24_LE format instead of mode)

francesco-vannini commented:
Hi @HiassofT does this mean that now samples with 16/24/32 can all be played without resampling? Also what do I need to do to test this particular commit on my Raspbian?

HiassofT commented:
24bit data wasn't actually resampled but padded to 32bit by the I2S controller.

I did some tests and it looks like this padding doesn't simply add zero-bits or repeat the LSB but the MSB of the next channel data is put on the I2S data line. In other words, the samples were padded with some semi-random, fixed value. Although that's more or less an academic issue it's still odd to have it and purists may prefer that their 24bit data is transmitted unmodified as exactly 24bit to the DAC - like it was implemented in the upstream driver.

This is also fixed by this PR, so simply speaking it makes sure all of your 16/24/32 bit data gets transmitted without any "conversion" as 16/24/32bit data to the DAC.

As for testing: it'd be easiest if you just download a tarball of my PR branch via this URL: https://github.com/HiassofT/rpi-linux/archive/4.14-pcm512x-24bit-fix.tar.gz and then compile a kernel from it as usual. This branch is based on the current 4.14 rpi branch, so basically you'll get a `BRANCH=next rpi-update` kernel plus the pcm512x 24bit fix.

francesco-vannini commented:
@HiassofT thanks for the explanation, I never realised this was the case.

I wouldn't want to be too much of a needy newbie but I don't do kernel compilation so regularly (once every three years maybe :) would you happen to have a couple of notes or a link on how to go about it?

Thanks a lot for this

pelwell commented:
Allow me: https://www.raspberrypi.org/documentation/linux/kernel/building.md

francesco-vannini commented:
The hugely famous RTFM or Google it up right? :) Thanks I had found that but wasn't sure. Keep you posted, get everyone off the internet in the meantime I might come up with a new deadly virus by mistake :))

HiassofT commented:
Ah, Phil was quicker than me :)

To make testing even easier I've uploaded a precompiled RPi2/3 kernel here: http://www.horus.com/~hias/tmp/rpi/rpi2-4.14-pcm512x-24bit-fix.tar.gz

Just download it and then extract it with
```
sudo tar zxf rpi2-4.14-pcm512x-24bit-fix.tar.gz -C /
```

francesco-vannini commented:
The internet is a safer place, thanks guys! This includes even #2364 ? It doesn't just realised.

miquel83blauw commented:
Tested it here with the LOCO-V2, everything works flawless!

pelwell commented:
Thanks. I'll merge this PR once I have a few more confirmations, or 16th February if there are no objections - two weeks is long enough.

iqaudio commented:
Phil, I will not get a chance to try but as the code was tested using our boards I will have to assume it’s fine. Sorry.

Gordon


> On 8 Feb 2018, at 15:27, Phil Elwell <notifications@github.com> wrote:
> 
> Thanks. I'll merge this PR once I have a few more confirmations, or 16th February if there are no objections - two weeks is long enough.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

francesco-vannini commented:
Tested a few samples with various samplerate and bitspersample and all seems fine for both #2365 and #2364 
Don't have a good enough DAC to handle 192KHz 24bit samples but at least aplay and omxplayer don't seem to complain.
I even took it upon myself to give it a go with a cross compilation of the branch just to see if I manage :)
Many thanks for your help guys.

francesco-vannini commented:
Guys I have tried the cross compilation, all seemed fine but when I run lsmod it returns an empty list.
These are the steps I followed:
```wget https://github.com/HiassofT/rpi-linux/archive/4.14-pcm512x-24bit-fix.tar.gz
tar xfvz 4.14-pcm512x-24bit-fix.tar.gz
git clone https://github.com/raspberrypi/tools ~/tools
echo PATH=\$PATH:~/tools/arm-bcm2708/gcc-linaro-arm-linux-gnueabihf-raspbian-x64/bin >> ~/.bashrc
source ~/.bashrc

cd linux
KERNEL=kernel7
make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- bcm2709_defconfig
make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- zImage modules dtbs
```

then moved the files to the Pi. Am I very far?

pelwell commented:
Assuming that you have copied the files to the right place, you are just missing the step that builds the aliases database (and a few other things) that allows the kernel to locate a module by compatible string, USB ID etc. You get this for free if you use the modules_install feature of the kernel Makefile, otherwise you must run `depmod` manually.

Unless you are running the same kernel version as the modules you want to re-index, which you aren't, you need to provide the kernel version number, for example:
```
sudo depmod 4.14.17-v7+
```
If the modules are not in `/lib/modules/4.14.17-v7+`, for example if you are preparing a card other than the one you are running from, then you'll also need to specify the base directory:
```
sudo depmod -b /mnt 4.14.17-v7+
```
(assuming the target filesystem is mounted on `/mnt`) which will index the modules in `/mnt/lib/modules/4.14.17-v7+`.

The equivalent command using the kernel Makefiles to also copy the modules would be:
```
sudo make ARCH=arm CROSS_COMPILE=arm-linux-gnueabihf- INSTALL_MOD_PATH=/mnt modules_install
```

francesco-vannini commented:
@pelwell ... several hours later ... it actually worked! Thanks, I've learnt a bit more today!

pelwell commented:
In the absence of any negative feedback I'm going to merge this PR. The bleeding-edge LibreELEC community will probably give it a good testing, and we can always revert or patch if there are problems.

adiadore commented:
hi guys, i've got an issue with S24_LE mode with my hifiberry dac+ pro
small buffering after 3-5 minutes and then followed by sound dropouts with video stuttering
i provided more infos [log](https://forum.kodi.tv/showthread.php?tid=298461&pid=2706220#pid2706220) and [dmesg ++](https://forum.kodi.tv/showthread.php?tid=298461&pid=2706278#pid2706278)
@hifiberry

HiassofT commented:
It looks like the sound stuttering issue in Kodi is caused by the Hifiberry DAC+ Pro not having suitable oscillators for 24bit mode.

In master mode the clock is derived  directly from the oscillator and with the 24.576MHz oscillator 48kHz 24bit would need a divisor of 10.666. As only integer divisors are possible the driver uses 10 and the effective rate is 51.2kHz instead of 48kHz. So audio runs too fast and Kodi can't sync that with video.

A simple fix for this is to disable 24bit mode, Kodi (and other programs) will then automatically use 32bit mode instead. This is basically the same as before the merge (24bit was transmitted with 32 clock cycles), only difference is that now userspace also knows about it.

I've prepared the changes and asked @MilhouseVH to include them in tomorrow's testbuild. @adiadore please test and report back if it works, then I'll PR the changes to the RPi kernel tree.

Preliminary (completely untested) changes are here if someone else wants to test: https://github.com/HiassofT/rpi-linux/commits/4.14-fix-pcm512x-master

Note: Allo Boss in master mode is quite certainly effected by this issue, too (it uses 49.152/45.158MHz oscillators), the other pcm512x based sound cards run pcm512x in slave mode and are fine.

hifiberry commented:
@HiassofT Sorry I wasn't able to test this in the last weeks. 
Is there no way to use the S24LE format with 32bit with per channel? Even padding with random data would be ok as they are not really relevant as they will be outside of the dynamic range of the DAC.

pelwell commented:
Unless we can find a fix for 24-bit data on the Hifiberry DAC+ Pro I'll have to revert this PR.

HiassofT commented:
@hifiberry could you please test this tree: https://github.com/HiassofT/rpi-linux/tree/4.14-pcm512x-set-tdm-slot

I've implemented .set_tdm_slot in the pcm512x driver so sound card drivers can optionally configure the number of bclk cycles. I also changed the hifiberry_dacplus and allo-boss-dac drivers to use this, so 16, 24 and 32bit should work as before.

@adiadore I'll ask milhouse to include this change in his next build instead of the previous one

hifiberry commented:
@HiassofT : Thank you very much. I'll have a look, but it might take some days to find time for it. Im just in the process of moving the workplace and it's a bit chaotic at the moment.

PeterPablo commented:
I can second that the firmware https://github.com/raspberrypi/firmware/commit/9ab68bf218ffc84423a2591658260676d373259a causes issues during playback (radio via kodi) on a raspbian. The previous firmware (https://github.com/Hexxeh/rpi-firmware/commit/98b90b75aa4352c148f2726c84d6f466534a89fc) works fine. I also have a HifiBerry DAC+ Pro. I can test the next firmware being available via `rpi-update`.

pelwell commented:
We've had some confirmation from @adiadore via the LibreElec forum that these patches fix the issue for him, so they've been cherry-picked into rpi-4.14.y to hit tonight's firmware build.

HiassofT commented:
@pelwell thanks, I was just about to create a PR for the fix :)

PeterPablo commented:
I confirm that on https://github.com/Hexxeh/rpi-firmware/commit/ab90bf1971e4b6cd8c75d6ff7c20bf08166b613e the issue is fixed: `Linux rpi2 4.14.21-v7+ #1095 SMP Fri Feb 23 18:21:29 GMT 2018 armv7l GNU/Linux`
Thank you!

pelwell commented:
Cool.

I've just noticed that I referred to @adiadore as "him" with no evidence to support that assumption - apologies if that is not the case.

PeterPablo commented:
For what it's worth: adiadore commented in the LibreELEC thread (https://forum.kodi.tv/showthread.php?tid=298461&pid=2706880#pid2706880) that his issue is resolved with the patch of @HiassofT (https://raw.githubusercontent.com/HiassofT/LibreELEC.tv/aa2746c76992c60087706c1bacd0073fd8d24f42/projects/RPi/patches/linux/linux-10-fix-pcm512x-master-mode.patch) applied. So all good it seems, verified by two people independently.
